iwyu_include_picker.cc

```diff
@@ -1329,13 +1329,32 @@ vector<string> IncludePicker::GetCandidateHeadersForFilepathIncludedFrom(
   // ConvertToQuotedInclude because it avoids trouble when the same
   // file is accessible via different include search-paths, or is
   // accessed via a symlink.
-  const string& quoted_include_as_typed
-      = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
-  if (!quoted_include_as_typed.empty()) {
-    vector<string>::iterator it = std::find(retval.begin(), retval.end(),
-                                            quoted_includee);
-    if (it != retval.end())
-      *it = quoted_include_as_typed;
+  vector<string>::iterator it = std::find(retval.begin(), retval.end(),
+                                          quoted_includee);
+  if (it != retval.end()) {
+      const string& quoted_include_as_typed
+          = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
+      if (!quoted_include_as_typed.empty()) {
+          *it = quoted_include_as_typed;
+      } else {
+          string strippedIncludee = included_filepath;
+          // Directory of the file is implicitly in the search path.
+          // So if we have the full path, shorten it
+          const string implicitHeaderSearchPath =
+                  CanonicalizeHeaderSearchPath(GetParentPath(including_filepath));
+          if (StripLeft(&strippedIncludee, implicitHeaderSearchPath)) {
+              VERRS(6) << "Adding include in same directory for "
+                       << included_filepath << ": " << strippedIncludee
+                       << "\n";
+              // If the included path is absolute, we prefer the short one
+              // else we prefer the current one, which is probably relative to
+              // a search directory
+              if(IsAbsolutePath(StripQuotes(*it)))
+                  retval.insert(it, "\"" + strippedIncludee + "\"");
+              else
+                  retval.push_back("\"" + strippedIncludee + "\"");
+          }
+      }
```

Comment:
This `else` branch remains the most confusing part for me. It is unclear why we do these actions only when `MaybeGetIncludeNameAsWritten` doesn't return anything. In what cases `MaybeGetIncludeNameAsWritten` returns empty string? It seems it's empty because of deficiencies in other parts of the program, not that it is empty on purpose in specific cases.

Also I tried to comment out this `else` block and see how tests behave. Tests broke in a way that IWYU recommends to use
```c++
#include "/subfolder/indirect_subfolder.h"
#include "/indirect.h"
```
To me this method doesn't look like appropriate place to fix leading slash. And it indicates that in some place we didn't strip path prefix correctly and inadvertently created false absolute path. If you know why that happened, it would be very helpful. Otherwise I'll need to figure out in debugger where those paths come from.